userd/helpers.go

```diff
@@ -0,0 +1,83 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package userd
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/godbus/dbus"
+
+	"github.com/snapcore/snapd/dirs"
+)
+
+var snapFromSender = snapFromSenderImpl
+
+func snapFromSenderImpl(conn *dbus.Conn, sender dbus.Sender) (string, error) {
+	pid, err := connectionPid(conn, sender)
+	if err != nil {
+		return "", fmt.Errorf("cannot get connection pid: %v", err)
+	}
+	snap, err := snapFromPid(pid)
+	if err != nil {
+		return "", fmt.Errorf("cannot find snap for connection: %v", err)
+	}
+	return snap, nil
+}
+
+func connectionPid(conn *dbus.Conn, sender dbus.Sender) (pid int, err error) {
+	call := conn.BusObject().Call("org.freedesktop.DBus.GetConnectionUnixProcessID", 0, sender)
+	if call.Err != nil {
+		return 0, call.Err
+	}
+	call.Store(&pid)
+	return pid, nil
+}
+
+// FIXME: move to osutil?
+func snapFromPid(pid int) (string, error) {
+	// racy :( - maybe move to use libapparmor but that won't work on
+	// every distro yet
+	m, err := filepath.Glob(filepath.Join(dirs.FreezerCgroupDir, "snap.*"))
```

Comment:
This should be avaialble in proc under `/proc/<pid>/cgroup`. For example for `snap run --shell hello-world`:
```
$ cat /proc/5873/cgroup 
10:devices:/user.slice
9:cpuset:/
8:net_cls,net_prio:/
7:freezer:/snap.hello-world
6:perf_event:/
5:pids:/user.slice/user-1000.slice/user@1000.service
4:cpu,cpuacct:/
3:memory:/
2:blkio:/
1:name=systemd:/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service
0::/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service
```

Reply:
Thanks, I have a look if it this is easier to parse than the current (indirect) appraoch.

Reply:
I did this now, its slightly shorter and for bonus points I added a test